CORS "Internal exception in JAX-RS processing" with "null" Origin

## Environment Details
* Helidon Version: 4.0.6
* Helidon MP
* JDK version: Amazon Coretto 21.0.2 Hotspot
* OS: Windows 11

----------

I justed enabled CORS in the quickstart maven example ( add `helidon-microprofile-cors` dependency and set `cors.enabled=true`) ,then I opened a new Firefox window, opened a console and typed `fetch("http://localhost:8080/"`. This lead to a CORS security warning `Content-Security-Policy: .... ("connect-src")`. 

I wanted to take a look at the headers hence I copied the network request via Firefox into a curl command and executed it:
```
curl 'http://172.18.32.1:8080/' -H 'User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:123.0) Gecko/20100101 Firefox/123.0' -H 'Accept: */*' -H 'Accept-Language: de,en-US;q=0.7,en;q=0.3' -H 'Accept-Encoding: gzip, deflate, br' -H 'Origin: null'
```

This returns a `Internal exception in JAX-RS processing` and a lengthy exception in the server logs ending with:
```
Caused by: java.lang.StringIndexOutOfBoundsException: Range [0, -1) out of bounds for length 4
	at java.base/jdk.internal.util.Preconditions$1.apply(Preconditions.java:55)
	at java.base/jdk.internal.util.Preconditions$1.apply(Preconditions.java:52)
	at java.base/jdk.internal.util.Preconditions$4.apply(Preconditions.java:213)
	at java.base/jdk.internal.util.Preconditions$4.apply(Preconditions.java:210)
	at java.base/jdk.internal.util.Preconditions.outOfBounds(Preconditions.java:98)
	at java.base/jdk.internal.util.Preconditions.outOfBoundsCheckFromToIndex(Preconditions.java:112)
	at java.base/jdk.internal.util.Preconditions.checkFromToIndex(Preconditions.java:349)
	at java.base/java.lang.String.checkBoundsBeginEnd(String.java:4865)
	at java.base/java.lang.String.substring(String.java:2834)
	at io.helidon.cors.CorsSupportHelper.originLocation(CorsSupportHelper.java:425)
	at io.helidon.cors.CorsSupportHelper$RequestTypeInfo.create(CorsSupportHelper.java:390)
	at io.helidon.cors.CorsSupportHelper.requestType(CorsSupportHelper.java:375)
```

I hence would argue that the CorsSupportHelper::originLocation should be more robust to not cause these exceptions.

--- a/cors/src/main/java/io/helidon/cors/CorsSupportHelper.java
+++ b/cors/src/main/java/io/helidon/cors/CorsSupportHelper.java
@@ -61,7 +61,10 @@ public class CorsSupportHelper<Q, R> {
 
     static final Logger LOGGER = Logger.getLogger(CorsSupportHelper.class.getName());
 
+    static final String OPAQUE_ORIGIN = "null"; // browsers might send this as Origin header if origin info is untrusted
+
     private static final Supplier<Optional<CrossOriginConfig>> EMPTY_SECONDARY_SUPPLIER = Optional::empty;
+
 
     private final String name;
 
@@ -392,7 +395,7 @@ public class CorsSupportHelper<Q, R> {
 
             return new RequestTypeInfo(originLocation,
                                        hostLocation,
-                                       originLocation.equals(hostLocation));
+                                       originLocation.equals(hostLocation) || originLocation.equals(OPAQUE_ORIGIN));
         }
     }
 
@@ -403,6 +406,9 @@ public class CorsSupportHelper<Q, R> {
         // Fast decision if there is no Origin header.
         if (originOpt.isEmpty()) {
             LogHelper.logIsRequestTypeNormalNoOrigin(silent, requestAdapter);
+            return true;
+        } else if (originOpt.get().equals(OPAQUE_ORIGIN)) {
+            LogHelper.logIsRequestTypeNormalOpaqueOrigin(silent, requestAdapter);
             return true;
         }
 
@@ -421,7 +427,7 @@ public class CorsSupportHelper<Q, R> {
         int originLastColon = origin.lastIndexOf(':');
 
         return origin + (
-                (originEndOfScheme == originLastColon)
+                (originEndOfScheme == originLastColon && originEndOfScheme >= 0)
                         ? ":" + portForScheme(origin.substring(0, originEndOfScheme))
                         : "");
     }

--- a/cors/src/main/java/io/helidon/cors/CorsSupportHelper.java
+++ b/cors/src/main/java/io/helidon/cors/CorsSupportHelper.java
@@ -407,9 +407,10 @@ public class CorsSupportHelper<Q, R> {
         if (originOpt.isEmpty()) {
             LogHelper.logIsRequestTypeNormalNoOrigin(silent, requestAdapter);
             return true;
-        } else if (originOpt.get().equals(OPAQUE_ORIGIN)) {
-            LogHelper.logIsRequestTypeNormalOpaqueOrigin(silent, requestAdapter);
-            return true;
+        }
+        if (isOriginOpaque(originOpt.get())) {
+            LogHelper.logOpaqueOrigin(silent, requestAdapter);
+            // Do not return. Continue processing having noted the opaque origin.
         }
 
         RequestTypeInfo result = requestType(originOpt.get(), requestAdapter.requestedUri());
@@ -420,6 +421,10 @@ public class CorsSupportHelper<Q, R> {
                                          result.originLocation,
                                          result.hostLocation);
         return result.isNormal;
+    }
+
+    static boolean isOriginOpaque(String origin) {
+        return origin.equals(OPAQUE_ORIGIN);
     }
 
     private static String originLocation(String origin) {

--- a/cors/src/main/java/io/helidon/cors/LogHelper.java
+++ b/cors/src/main/java/io/helidon/cors/LogHelper.java
@@ -79,6 +79,16 @@ class LogHelper {
                                                    List.of("header " + HeaderNames.ORIGIN + " is absent")));
     }
 
+    static <T> void logIsRequestTypeNormalOpaqueOrigin(boolean silent, CorsRequestAdapter<T> requestAdapter) {
+        if (silent || !CorsSupportHelper.LOGGER.isLoggable(DECISION_LEVEL)) {
+            return;
+        }
+        CorsSupportHelper.LOGGER.log(DECISION_LEVEL,
+                                     String.format("Request %s is not cross-host: %s",
+                                                   requestAdapter,
+                                                   List.of("header " + HeaderNames.ORIGIN + " value is opaque")));
+    }
+
     static <T> void logIsRequestTypeNormal(boolean result, boolean silent, CorsRequestAdapter<T> requestAdapter,
             Optional<String> originOpt, String effectiveOrigin, String host) {
         if (silent || !CorsSupportHelper.LOGGER.isLoggable(DECISION_LEVEL)) {

--- a/cors/src/main/java/io/helidon/cors/LogHelper.java
+++ b/cors/src/main/java/io/helidon/cors/LogHelper.java
@@ -79,14 +79,14 @@ class LogHelper {
                                                    List.of("header " + HeaderNames.ORIGIN + " is absent")));
     }
 
-    static <T> void logIsRequestTypeNormalOpaqueOrigin(boolean silent, CorsRequestAdapter<T> requestAdapter) {
+    static <T> void logOpaqueOrigin(boolean silent, CorsRequestAdapter<T> requestAdapter) {
         if (silent || !CorsSupportHelper.LOGGER.isLoggable(DECISION_LEVEL)) {
             return;
         }
         CorsSupportHelper.LOGGER.log(DECISION_LEVEL,
-                                     String.format("Request %s is not cross-host: %s",
+                                     String.format("Request %s specifies opaque origin: %s",
                                                    requestAdapter,
-                                                   List.of("header " + HeaderNames.ORIGIN + " value is opaque")));
+                                                   List.of("header " + HeaderNames.ORIGIN)));
     }
 
     static <T> void logIsRequestTypeNormal(boolean result, boolean silent, CorsRequestAdapter<T> requestAdapter,

--- a/cors/src/test/java/io/helidon/cors/CorsSupportHelperTest.java
+++ b/cors/src/test/java/io/helidon/cors/CorsSupportHelperTest.java
@@ -76,6 +76,14 @@ class CorsSupportHelperTest {
     }
 
     @Test
+    void opaqueOrigin() {
+        assertThat("Opaque origin",
+                   CorsSupportHelper.requestType(CorsSupportHelper.OPAQUE_ORIGIN,
+                                                 uriInfo("ok.com", 8010, false)).isNormal(),
+                   is(true));
+    }
+
+    @Test
     void sameNodeSamePort() {
         assertThat("Default origin port",
                    CorsSupportHelper.requestType("http://ok.com",

--- a/cors/src/test/java/io/helidon/cors/CorsSupportHelperTest.java
+++ b/cors/src/test/java/io/helidon/cors/CorsSupportHelperTest.java
@@ -78,9 +78,12 @@ class CorsSupportHelperTest {
     @Test
     void opaqueOrigin() {
         assertThat("Opaque origin",
-                   CorsSupportHelper.requestType(CorsSupportHelper.OPAQUE_ORIGIN,
-                                                 uriInfo("ok.com", 8010, false)).isNormal(),
+                   CorsSupportHelper.isOriginOpaque(CorsSupportHelper.OPAQUE_ORIGIN),
                    is(true));
+
+        assertThat("Non-opaque origin",
+                   CorsSupportHelper.isOriginOpaque("http://ok.com"),
+                   is(false));
     }
 
     @Test

